### ISSUE TITLE
[Tracking Issue] 3.0 Release

### ISSUE THREAD

**Issue by Nihlus** [opentk/opentk#702]
This issue tracks the issues and pull requests we'd like to get into the final 3.0 release. This list is not yet finalized and is subject to change, but can be considered a checklist for things that need to happen before the 3.0 release.

If you have an issue or pull request that you'd like to be added to this list, please discuss it below. In general, we need you to be prepared to work on the issue yourself to add it to the list, unless it is a minor change.

We need as much help as we can get with all of these issues - testing on different platforms, quality assurance, etc. If you have some spare time, please chip in.

Onwards towards 3.0!

## Issues
- [ ] #678 Quaternion * Vector3 is not implemented for double precision
- [ ] #669 Modifier keys not detected as individual keypresses on macOS
- [ ] #668 NativeWindow.CursorVisible reseting mouse position to 0/0
- [x] #710 "Failed to load assembly" in Rewrite.exe when trying to build
- [x] #696 Mouse snaps back to window bounds when window loses focus
- [x] #642 Rewrite.exe is adding spurious references to .NET 4.0 assemblies to OpenTK.dll
- [x] #688 Missing enums for anisotropic texture filtering
- [x] #656 Crash on Mouse.GetState() when mouse is unplugged on MacOS

## Pull Requests
- [ ] #679 Make Vector3d and Vector3 equal in interface
- [ ] #589 [OSX test required]Separate confining and visiblility for mouse cursor
- [x] #692 WIP: Quaternion
- [x] #697 Use custom cecil assembly resolver
- [x] #700 Conditional stylecop activation
- [x] #703 Remove bindings from Git
- [x] #698 Update bindings from OpenGL-Registry

**Comment by ThomasD13** [comment]
I think you can remove #678 from the list. It will be solved with #679.

**Comment by VPeruS** [comment]
It should autocheck entry on merge

**Comment by Nihlus** [comment]
They're both part of the milestone (the issue that tracks it, and the PR that fixes it), so I'd say it's fine to have both.

**Comment by ThomasD13** [comment]
#692 was closed by me by accident. I worked further on it at #714.

**Comment by djohns451** [comment]
#733 would be great to have with a major point release.

**Comment by Nihlus** [comment]
Agreed - one of the documentation proposals should definitely be a part of it. I'll bring it up.

**Comment by Nihlus** [comment]
By general vote, we've decided to move forward with the canonical 3.0 release. Please be aware that OpenTK will enter a code freeze period on the 1st of May, whereafter no new pull requests will be accepted or merged until the final 3.0 release. All outstanding pull requests have until then to be finalized or rejected.

On the date of the code freeze, the `develop` branch will be cloned into a `3.0-release` branch, which will be kept around. The maintainer team will not actively backport fixes or features to this branch, but the community is free to make pull requests against it, which may result in point releases down the line.

This release will, as mentioned, be the final release that targets the full .NET framework. Future releases will target the .NET Standard, and will take a more modular approach.

**Comment by tzachshabtay** [comment]
I went over all existing PRs (it would be a shame to lose the work that was already done).

From what I can tell (please correct me if I'm wrong):

Good to merge: #719, #737 (the build fails here, but from what I can tell it's not related to the PR itself? And can't see the actual error in the build logs, and all tests pass)
Waiting for @VPeruS: #589 (not always working as expected?), #679 (missing tests?), #731 (not always working as expected?)
Waiting for @Frassle: #610 (waiting for feedback), #651 (not always working as expected?)
Waiting for @Jayrod246: #738 (missing tests)
Waiting for @SnowmanTackler: #744 (missing tests)
Waiting for @djohns451: #733 (still marked as WIP -> though I guess we maybe can keep this PR alive after 3.0 as it's not likely to conflict?)
Waiting for @tyronx: #666 (problems with design?)
Not sure: #728 (I guess can be rejected? as functionality deemed not useful)

**Comment by tyronx** [comment]
On #666, yes that push request should probably be made configurable to either have old or new behavior.  I have those changed merged to my fork that's used in my game for 2 years now so it's working fine for me. However it's unlikely that I will add a configuration option to that anytime soon, as I can just always apply the changes of that PR to my fork. Maybe maybe @copygirl feels like it working on it if she feels inspired.

**Comment by Frassle** [comment]
#610 is a big change and was very hard to test. I'm sure it will require bug fixes, so not sure we want that on a low-maint branch. It would be very good to have, keep it around and maybe get into 4.0 early.

#651 I thought might be a quick fix, it isn't. I don't have any controllers/joysticks to test with so unlikely I'll get it finished. Might even be best to reject it.

**Comment by Nihlus** [comment]
3.0 has now been release, and work on 4.0 has begun.

**Comment by tyronx** [comment]
Just for clarification @Nihlus, without #589 + peppy patch controlling a first person game (as in, locked and centered mouse cursor) on mac os is essentially impossible.

**Comment by Nihlus** [comment]
@tyronx Hopefully we can merge that into the first 4.0 release, then.

**Comment by djohns451** [comment]
Seems a bit strange for us to be defining a new major version per release. https://semver.org/

**Comment by winterhell** [comment]
Supposedly everyone is moving to .NET Core so there is no reason to support OpenTK 3.x any longer

**Comment by Nihlus** [comment]
@copygirl Generally a waste of time. The same fixes would need to be applied to the 4.0 branch, and that's where the work is going to happen now.

@djohns451 We are following semver - both the 3.0 release and the 4.0 release have been or will be breaking changes.

@winterhell .NET Standard, not neccesarily core.
